### PR TITLE
Pp 66 create edit delete lesson

### PR DESCRIPTION
**Pull request by mariyador**
### Description

The functionality for adding, editing, deleting lesson has been implemented.

!!!! To test this you need to be on PP-65-update-lessonscontroller-back branch (BACKEND)

### Screenshots
AddLesson form with new data:
Screenshot 2024-09-05 at 1 00 29 AM

Lesson created successfully:
Screenshot 2024-09-05 at 1 00 36 AM

Delete this lesson:
Screenshot 2024-09-05 at 1 00 41 AM

Edit the descriptions of Lesson 1:
Screenshot 2024-09-05 at 1 01 04 AM

### CODE REVIEW

File: src/components/pages/dashboard-pages/AddLesson.jsx [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        const filteredClasses = classes.filter(
+          (classItem) =>
+            classItem.createdBy === userData._id &&
+            classItem.classStudents?.some(
+              (student) => student.userId === selectedStudentId
+            )
+        );
+        setFilteredClasses(filteredClasses);
+      } catch (error) {
+        console.error('Error fetching classes:', error);
```
Comment (vrudnitskaya) [inline src/components/pages/dashboard-pages/AddLesson.jsx:38]: Will we show this error to the user?

File: src/components/pages/dashboard-pages/EditLesson.jsx [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+          lessonDescription: lessonInfo.lessonDescription || '',
+          type: lessonInfo.type || '',
+          hometask: lessonInfo.hometask || '',
+          lessonSchedule: {
+            date: formatDate(lessonInfo.lessonSchedule?.date),
+            startTime: lessonInfo.lessonSchedule?.startTime || '',
+          },
+        });
+      } catch (error) {
+        console.error('Error fetching lesson:', error);
```
Comment (vrudnitskaya) [inline src/components/pages/dashboard-pages/EditLesson.jsx:43]: I think the user should be informed if something went wrong

File: src/components/pages/dashboard-pages/EditLesson.jsx [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+  const handleSubmit = async (e) => {
+    e.preventDefault();
+    setIsLoading(true);
+    const token = userData.token;
+
+    try {
+      await editLesson(token, formData, selectedStudentId, lessonId);
+      navigate('/dashboard/students'); 
+    } catch (error) {
+      setFormErrors(error);
```
Comment (vrudnitskaya) [inline src/components/pages/dashboard-pages/EditLesson.jsx:93]: all your settings of the form error have different approaches.

File: src/components/portfolio/FormUploadPortfolio.jsx [hunk 4]
```diff
@@ -0,0 +1,33 @@
+import { useEffect, useState } from "react";
```
Comment (vrudnitskaya) [inline src/components/portfolio/FormUploadPortfolio.jsx:1]: you have the old version of files from my pr, please check before merging

File: src/components/pages/dashboard-pages/EditLesson.jsx [hunk 5]
```diff
@@ -0,0 +1,110 @@
+import React, { useState, useEffect } from 'react';
```
Comment (JoHeyJo) [inline src/components/pages/dashboard-pages/EditLesson.jsx:1]: Stretch goal: If you have time I would add a simple message that can be displayed to the user of required fields. e.g. if you try to edit without a title nothing happens. There is no indication of why the form is not being submitted.

File: src/components/pages/dashboard-pages/EditLesson.jsx [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    lessonDescription: '',
+    type: '',
+    hometask: '',
+    lessonSchedule: { date: '', startTime: '' },
+  });
+  const [formErrors, setFormErrors] = useState({});
+  const [isLoading, setIsLoading] = useState(false);
+
+  useEffect(() => {
+    const fetchClasses = async () => {
```
Comment (JoHeyJo) [inline src/components/pages/dashboard-pages/EditLesson.jsx:25]: The name is a little misleading b/c we are actually fetching lesson details and returning lesson info.

File: src/components/pages/dashboard-pages/EditLessonForm.jsx [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+          <div className="flex gap-4 my-4">
+            <p className="w-1/4">Type:</p>
+            <div className="flex flex-col gap-2 w-3/4">
+              <label className="flex gap-1 items-center">
+                <input
+                  type="radio"
+                  name="type"
+                  value="online"
+                  checked={formData.type === 'online'}
+                  className="w-4 h-4 accent-lightGreen focus:darkGreen"
```
Comment (JoHeyJo) [inline src/components/pages/dashboard-pages/EditLessonForm.jsx:85]: nit picky: To clean things up you can extract these repetitive classes into a constant and pass that along. Keep this in mind for future code bases to clean things up.

File: src/components/pages/dashboard-pages/LessonForm.jsx [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+                <input
+                  type="radio"
+                  name="type"
+                  value="online"
+                  className="w-4 h-4 accent-lightGreen focus:darkGreen"
+                  onChange={handleChange}
+                />
+                Online
+              </label>
+              <label className="flex gap-1 items-center">
```
Comment (JoHeyJo) [inline src/components/pages/dashboard-pages/LessonForm.jsx:107]: label is missing a for attribute to associate it with its corresponding input.

File: src/components/pages/dashboard-pages/Students.jsx [hunk 9]
```diff
@@ -4,6 +4,7 @@ import {
   getAllUsersInfo,
```
Comment (JoHeyJo) [inline src/components/pages/dashboard-pages/Students.jsx:4]: This component is working just fine. So let's not worry about refactoring. However, there is a lot going on. For future codebases just keep in mind to keep concerns separated. Also, a name like `TeacherStudent` is ambiguous so it makes it even more difficult as a developer to decide what logic belongs inside.

**Review by JoHeyJo — CHANGES_REQUESTED** [review]
Nice job implementing this. It's all working very smoothly. A few things to consider to polish up the app.

File: src/components/pages/dashboard-pages/EditLesson.jsx (hunk 6 above)
Comment (mariyador) [reply src/components/pages/dashboard-pages/EditLesson.jsx:25]: Fixed

File: src/components/pages/dashboard-pages/EditLesson.jsx (hunk 5 above)
Comment (mariyador) [reply src/components/pages/dashboard-pages/EditLesson.jsx:1]: Thank you! Done)

File: src/components/pages/dashboard-pages/LessonForm.jsx (hunk 8 above)
Comment (mariyador) [reply src/components/pages/dashboard-pages/LessonForm.jsx:107]: Fixed

File: src/components/pages/dashboard-pages/Students.jsx (hunk 9 above)
Comment (mariyador) [reply src/components/pages/dashboard-pages/Students.jsx:4]: Thank you for the feedback! I’ll keep this in mind for future work and will aim to better organize the logic and choose more descriptive names for components.

File: src/components/portfolio/FormUploadPortfolio.jsx (hunk 4 above)
Comment (mariyador) [reply src/components/portfolio/FormUploadPortfolio.jsx]: Thank you, Valya!
Already pull a new dev

File: src/components/pages/dashboard-pages/AddLesson.jsx (hunk 1 above)
Comment (mariyador) [reply src/components/pages/dashboard-pages/AddLesson.jsx:38]: Fixed

File: src/components/pages/dashboard-pages/EditLesson.jsx (hunk 2 above)
Comment (mariyador) [reply src/components/pages/dashboard-pages/EditLesson.jsx:47]: Fixed

**Review by JoHeyJo — APPROVED** [review]
Great job implementing all this new functionality! Just keep in mind the notes that address best practices going forward to really help clean up the code base and make it more navigable.

**Review by vrudnitskaya — APPROVED** [review]
Great job, Mariya!

**Pull request merged**